lib/Conversion/CalyxToFSM/CalyxToFSM.cpp

```diff
@@ -268,6 +268,71 @@ LogicalResult CompileFSMVisitor::visit(StateOp currentState, EnableOp enableOp,
   return success();
 }
 
+// CompileInvoke is used to convert invoke operations to group operations and
+// enable operations.
+class CompileInvoke {
+public:
+  CompileInvoke(ComponentOp component, OpBuilder builder)
+      : component(component), builder(builder) {}
+  void compile();
+
+private:
+  void lowerInvokeOp(InvokeOp invokeOp);
+  ComponentOp component;
+  OpBuilder builder;
+  // The groupNameMap is used to help generate the names of calyx.group,
+  // for example, if key is reg, value is 0, then calyx.group is named
+  // calyx.group invoke_reg_0. When an invoke is encountered, if the
+  // callee is not in the groupNameMap, then the callee is added to
+  // the groupNameMap and the corresponding value is made to be 0.
+  llvm::StringMap<size_t> groupNameMap;
+};
+
+// Access all invokeOp.
+void CompileInvoke::compile() {
+  llvm::SmallVector<InvokeOp> invokeOps =
+      component.getControlOp().getInvokeOps();
+  for (InvokeOp op : invokeOps) {
+    lowerInvokeOp(op);
+  }
+}
+
+// Convert an invoke operation to a group operation and an enable operation.
+void CompileInvoke::lowerInvokeOp(InvokeOp invokeOp) {
+  // Create a ConstantOp to assign a value to the go port.
+  Operation *prevNode = component.getWiresOp().getOperation()->getPrevNode();
+  builder.setInsertionPointAfter(prevNode);
+  hw::ConstantOp constantOp = builder.create<hw::ConstantOp>(
+      prevNode->getLoc(), builder.getI1Type(), 1);
+  Location loc = component.getWiresOp().getLoc();
+  // Set the insertion point at the end of the wires block.
+  builder.setInsertionPointToEnd(component.getWiresOp().getBodyBlock());
+  llvm::StringRef callee = invokeOp.getCallee();
+  std::string groupNameHead = "invoke_" + callee.str() + "_";
+  size_t &groupNameTail = groupNameMap[callee];
+  std::string groupName = groupNameHead + std::to_string(groupNameTail);
+  // Check if the group name is correct.
+  while (component.getWiresOp().lookupSymbol(groupName))
```

Comment:
I think this can probably be simplified. It looks like the number in `groupNameMap[callee]` is not re-used for future callee's of the same name, right? It seems like if the callee has already had a GroupOp generated, this will loop until there is a fresh groupNameTail counter for that callee name.

If I understand it correctly, I'd drop the groupNameMap completely, and replace it with a counter that just increments each time lowerInvokeOp is called, and appended to the groupName string.

Reply:
I think you should be able to remove this loop completely. What I'm suggesting is to make `groupNameTail` an instance variable on `CompileInvoke`, remove this loop, and always just do `groupName = groupNameHead + std::to_string(++groupNameTail);`. That way you aren't doing a loop and symbol lookups at all--you know each invoke you compile gets a distinct group name.

Reply:
This only guarantees unique-ness amongst invoke transitions, not all transitions right?

Reply:
The issue would be if there was a group that was named similarly, found in the same control operator:

```
seq {
 group invoke_r0_0 { ... }
 invoke_r0 (...) -> (...) /* ...renamed to invoke_r0_0 */
}
```

Reply:
I guess it is possible a group name that was not generated from an invoke conflicts, but our compiler is making the group names, so I thought we could be sure the `invoke_` prefix is unique? It just seems like if we want to generate unique group names, we shouldn't have to check what group names have already been generated.

Reply:
These are "FSM transition" names, which is now the union of group names and invokes. So, we do need some additional logic.

I think the lookup is just a safety measure, and we shouldn't expect it to occur more than once in the majority of cases. Another hack could be to add some standard prefix to invoke transitions, e.g., `__invoke`.

Reply:
To tell you the truth, I think it would be better to use a Map to minimize the checking of previously generated group names.

Reply:
Can you elaborate? I think Mike's suggestion of a class-provided size_t is equivalent to using a map (in cases where we have multiple invokes)

Reply:
Origin of the name of the group,`compiler-generated`, ` user-defined`.
The reason for checking group names is that it is impossible to determine what group names the user has defined, so checking group names I think is necessary, and the use of map reduces the number of checks, which I think is also necessary.

Reply:
I see the need to loop and ensure the group name is not conflicting. I don't see what using a map gains you--you just need to be able to generate new strings in a loop until one does not conflict. In other words, I think the current implementation is fine.

Reply:
I think the current implementation is also good, but consider a situation where if the same component is called 10 times, then the truth of the first call should only need 1 lookup, the second should do 2 lookup(the generated name is repeatedly searched.), the third should do 3 loopup(2 generated name is repeatedly searched.), and then similar situations.
So I'm thinking about this, but if you use a map, the value of map holds the numeric portion of the name that should be generated this time, this reduces the number of unnecessary queries.If you do this, the query is mainly to determine whether the user used the name.

Reply:
Maybe I'm misunderstanding, but can't that be fixed using a class member 'index' instead of a mapping? That way, it always produces a unique name amongst 'invoke' operations.

I think it would be useful to add a test example that illustrate invoke/group name clash in any case.

Reply:
The counter that is suffixed to the generated group name is shared across invocations and incremented on every call right? Shouldn't that mean every time this is called, the generated group name should already be unique w.r.t. other calls? I think the only time there could be a conflict is if the generated name happens to conflict with a name derived from user code.

Reply:
Oh I see the issue. Part of what I suggested earlier hasn't been implemented: the `size_t groupNameTail = 0;` should be declared as a private member of the `CompileInvoke` class.

Reply:
The very first version is also what you think, that time `groupName = "invoke" + groupNameTail`. but later I consider the readability, later became `groupName = "invoke_" + callee + "_" + groupNameTail`, so use the map, the map of the key is used to identify the callee,value is groupNameTail,but later we went on to discuss the `ConstantOp` issue, readability here looks to be less important because if the final pass generates a result that is just a transition. I'll change it to the original form if you think it's necessary.

Reply:
I don't think you need to change it back to the original form, I think `size_t groupNameTail = 0;` should be declared as a private member of `CompileInvoke`, so it is shared between multiple calls to `compile`. This will avoid the issue where each call has to search further to find a unique name.

Reply:
I think you need to make a choice about which implementation to adopt here. Do we want readability or not?

Reply:
As I mentioned elsewhere, this is a compiler IR. The readability is only a concern to us, the compiler developers who have to debug errors and write test cases. The output in this PR is perfectly acceptable to me. I'm just suggesting to make the `groupNameTail` shared state across multiple calls to `compile`.

Reply:
I'm happy to discuss things with you here. Ok, I'll follow your suggestion.:smile: